replication-api-impl/src/main/java/com/connexta/replication/api/impl/queue/memory/MemorySiteQueue.java

```diff
@@ -0,0 +1,426 @@
+  private final List<Deque<MemoryTask>> pending = new ArrayList<>();
+
+  /** Ordered set of active tasks that are still locked and being worked on. */
+  private final Set<MemoryTask> active = new LinkedHashSet<>();
+
+  /** The capacity bound, or Integer.MAX_VALUE if none. */
+  private final int capacity;
+
+  /** Current number of tasks. */
+  private final AtomicInteger count;
+
+  /** Current number of pending tasks. */
+  private final AtomicInteger pendingCount;
+
+  /** Current number of active tasks. */
+  private final AtomicInteger activeCount;
+
+  /** Lock held by take, poll, etc. */
+  private final ReentrantLock takeLock = new ReentrantLock();
+
+  /** Wait queue for waiting takes. */
+  private final Condition notEmpty = takeLock.newCondition();
+
+  /** Lock held by put, offer, etc. */
+  private final ReentrantLock putLock = new ReentrantLock();
+
+  /** Wait queue for waiting puts. */
+  private final Condition notFull = putLock.newCondition();
+
+  private final MeterRegistry registry;
+
+  private final Counter queuedCounter;
+  private final Counter requeuedCounter;
+  private final Counter successCounter;
+  private final Counter failCounter;
+  private final Timer timer;
+  private final Timer pendingTimer;
+  private final Timer activeTimer;
+
+  /**
+   * Instantiates a default in-memory site queue.
+   *
+   * @param broker the broker for which we are instantiating this queue
+   * @param site the corresponding site id
+   * @param capacity the capacity of the queue this broker manages
+   * @param registry the micrometer registry to report metrics
+   */
+  public MemorySiteQueue(
+      MemoryQueueBroker broker, String site, int capacity, MeterRegistry registry) {
+    final Iterable<Tag> tags = Collections.singleton(Tag.of("site", site));
+
+    this.broker = broker;
+    this.site = site;
+    this.capacity = capacity;
+    this.registry = registry;
+    this.count = registry.gauge("replication_tasks_in_queue", tags, new AtomicInteger());
+    this.pendingCount =
+        registry.gauge("replication_pending_tasks_in_queue", tags, new AtomicInteger());
+    this.activeCount =
+        registry.gauge("replication_active_tasks_in_queue", tags, new AtomicInteger());
+    for (int i = MemorySiteQueue.MIN_PRIORITY; i <= MemorySiteQueue.MAX_PRIORITY; i++) {
+      pending.add(
+          registry.gaugeCollectionSize(
+              "replication_pending_priority_" + i + "_tasks_in_queue", tags, new LinkedList<>()));
+    }
+    this.queuedCounter = registry.counter("replication_queued_tasks", tags);
+    this.requeuedCounter = registry.counter("replication_requeued_tasks", tags);
+    this.successCounter = registry.counter("replication_success_tasks", tags);
+    this.failCounter = registry.counter("replication_fail_tasks", tags);
+    this.timer = registry.timer("replication_tasks_in_queue_timer", tags);
+    this.pendingTimer = registry.timer("replication_pending_tasks_in_queue_timer", tags);
```

Comment:
✏️ We shouldn't be ending the name with the meter type, but rather the unit the meter is tacking, so `tasks`.